effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -0,0 +1,549 @@
+    override final def tag = Tags.FlatMap
+  }
+
+  final case class Point[A](value: () => A) extends IO[A] {
+    override final def tag = Tags.Point
+  }
+
+  final case class Strict[A](value: A) extends IO[A] {
+    override final def tag = Tags.Strict
+  }
+
+  final case class SyncEffect[A](effect: () => A) extends IO[A] {
+    override final def tag = Tags.SyncEffect
+  }
+
+  final case class Fail[A](failure: Throwable) extends IO[A] {
+    override final def tag = Tags.Fail
+  }
+
+  final case class AsyncEffect[A](register: (Throwable \/ A => Unit) => AsyncReturn[A]) extends IO[A] {
+    override final def tag = Tags.AsyncEffect
+  }
+
+  final case class Attempt[A](value: IO[A]) extends IO[Throwable \/ A] {
+    override final def tag = Tags.Attempt
+  }
+
+  final case class Fork[A](value: IO[A], handler: Maybe[Throwable => IO[Unit]]) extends IO[Fiber[A]] {
+    override final def tag = Tags.Fork
+  }
+
+  final case class Race[A0, A1, A](left: IO[A0], right: IO[A1], finish: (A0, Fiber[A1]) \/ (A1, Fiber[A0]) => IO[A]) extends IO[A] {
+    override final def tag = Tags.Race
+  }
+
+  final case class Suspend[A](value: () => IO[A]) extends IO[A] {
+    override final def tag = Tags.Suspend
+  }
+
+  final case class Bracket[A, B](acquire: IO[A], release: (BracketResult[B], A) => IO[Unit], use: A => IO[B]) extends IO[B] {
+    override final def tag = Tags.Bracket
+  }
+
+  final case class Uninterruptible[A](io: IO[A]) extends IO[A] {
+    override final def tag = Tags.Uninterruptible
+  }
+
+  final case class Sleep(duration: Duration) extends IO[Unit] {
+    override final def tag = Tags.Sleep
+  }
+
+  final case class Supervise[A](value: IO[A], error: Throwable) extends IO[A] {
+    override final def tag = Tags.Supervise
+  }
+
+  /**
+   * Lifts a strictly evaluated value into the `IO` monad.
+   */
+  final def now[A](a: A): IO[A] = Strict(a)
+
+  /**
+   * Lifts a non-strictly evaluated value into the `IO` monad. Do not use this
+   * function to capture effectful code. The result is undefined but may
+   * include runtime errors.
+   */
+  final def point[A](a: => A): IO[A] = Point(() => a)
+
+  /**
+   * Raises the specified error. The moral equivalent of `throw` for pure code.
+   */
+  final def fail[A](t: Throwable): IO[A] = Fail(t)
+
+  /**
+   * Strictly-evaluated unit lifted into the `IO` monad.
+   */
+  final val unit: IO[Unit] = now(())
+
+  /**
+   * Sleeps for the specified duration. This is always asynchronous.
+   */
+  final def sleep(duration: Duration): IO[Unit] = Sleep(duration)
+
+  /**
+   * Supervises the specified action, which ensures that any actions directly
+   * forked by the action are killed with the specified error upon the action's
+   * own termination.
+   */
+  final def supervise[A](io: IO[A], error: Throwable): IO[A] = Supervise(io, error)
+
+  /**
+   * Flattens a nested action.
+   */
+  final def flatten[A](io: IO[IO[A]]): IO[A] = io.flatMap(a => a)
```

Comment:
Why there is no `flatten` method on `IO[A]`? Given `ev: A <:< IO[B]` it's just a `flatMap(ev)`. Same question about `absolve`.

Reply:
Because this should be `Liskov` but we do not have such a type class yet in Scalaz 8. 😄 I will definitely add both when we get `Liskov`.

Reply:
https://github.com/scalaz/scalaz/blob/series/8.0.x/base/shared/src/main/scala/scalaz/typeclass/Liskov.scala

Reply:
Ah, I see. Ok then :)